File: python/tests/test_transport_subjects_samples.py
"""Stage 6 / Task 6.3 (transport-spec v0.11 §4.22):
SUBJECT_METADATA (0x19) + SAMPLE_METADATA (0x1A) packets through
:class:`TransportWriter` + :class:`TransportReader`.

Wire layout per §4.22::

    arrow_ipc_length:    uint32
    arrow_ipc:           bytes[arrow_ipc_length]   # self-describing Arrow IPC

All multi-byte integers LITTLE-ENDIAN (spec §1.7).

Java parity: ``TransportSubjectsSamplesTest`` (commit ``dd211600``).
"""
from __future__ import annotations

import io
from pathlib import Path

from ttio.sample import Sample
from ttio.spectral_dataset import SpectralDataset
from ttio.subject import Subject
from ttio.transport.codec import TransportReader, TransportWriter
from ttio.transport.packets import PacketType


def _build_subjects_only(target: Path) -> Path:
    """Mirror Java's ``FixtureBuilder.buildSubjectsOnly``."""
    subjects = [
        Subject(
            external_id="S1",
            project="STUDY-A",
            sex="F",
            birth_year=1985,
            attributes={"site": "NYC"},
        ),
        Subject(
            external_id="S2",
            project="STUDY-A",
            sex="M",
            birth_year=1990,
            attributes={},
        ),
    ]
    SpectralDataset.write_minimal(
        target,
        title="subjects_only",
        isa_investigation_id="",
        runs={},
        subjects=subjects,
    )
    return target


def _build_samples_only(target: Path) -> Path:
    """Mirror Java's ``FixtureBuilder.buildSamplesOnly``."""
    samples = [
        Sample(
            sample_id="bio1",
            subject_external_id="",  # standalone sample
            sample_kind="tissue",
            collected_at=1700000000,
            attributes={"ph": "7.4"},
        ),
        Sample(
            sample_id="bio2",
            subject_external_id="",
            sample_kind="plasma",
            collected_at=0,  # sentinel
            attributes={},
        ),
    ]
    SpectralDataset.write_minimal(
        target,
        title="samples_only",
        isa_investigation_id="",
        runs={},
        samples=samples,
    )
    return target


def test_subjects_only_round_trip_through_writer_and_reader(
    tmp_path: Path,
) -> None:
    """Round-trip a subjects-only fixture through write_dataset +
    materialize and verify the rows survive."""
    src = _build_subjects_only(tmp_path / "src.tio")
    tis = tmp_path / "src.tis"
    rt = tmp_path / "rt.tio"

    with SpectralDataset.open(src) as ds:
        with open(tis, "wb") as out, TransportWriter(out) as w:
            w.write_dataset(ds)

    with TransportReader(tis) as r:
        materialised = r.read_to_dataset(output_path=rt)
        materialised.close()

    with SpectralDataset.open(src) as a, SpectralDataset.open(rt) as b:
        subs_a = a.subjects
        subs_b = b.subjects
        assert len(subs_a) == len(subs_b) == 2
        a_by_id = {s.external_id: s for s in subs_a}
        b_by_id = {s.external_id: s for s in subs_b}
        for ext_id in a_by_id:
            assert a_by_id[ext_id].project == b_by_id[ext_id].project
            assert a_by_id[ext_id].sex == b_by_id[ext_id].sex
            assert a_by_id[ext_id].birth_year == b_by_id[ext_id].birth_year
            assert a_by_id[ext_id].attributes == b_by_id[ext_id].attributes
        # Cross-cutting: samples must remain empty.
        assert a.samples == []
        assert b.samples == []


def test_samples_only_round_trip_through_writer_and_reader(
    tmp_path: Path,
) -> None:
    """Round-trip a samples-only fixture; sentinel-0 collected_at
    survives the Arrow null-encode/decode dance."""
    src = _build_samples_only(tmp_path / "src.tio")
    tis = tmp_path / "src.tis"
    rt = tmp_path / "rt.tio"

    with SpectralDataset.open(src) as ds:
        with open(tis, "wb") as out, TransportWriter(out) as w:
            w.write_dataset(ds)

    with TransportReader(tis) as r:
        materialised = r.read_to_dataset(output_path=rt)
        materialised.close()

    with SpectralDataset.open(src) as a, SpectralDataset.open(rt) as b:
        samp_a = a.samples
        samp_b = b.samples
        assert len(samp_a) == len(samp_b) == 2
        a_by_id = {s.sample_id: s for s in samp_a}
        b_by_id = {s.sample_id: s for s in samp_b}
        for sid in a_by_id:
            assert a_by_id[sid].subject_external_id == b_by_id[sid].subject_external_id
            assert a_by_id[sid].sample_kind == b_by_id[sid].sample_kind
            assert a_by_id[sid].collected_at == b_by_id[sid].collected_at
            assert a_by_id[sid].attributes == b_by_id[sid].attributes
        # bio2 sentinel collected_at == 0 must survive.
        assert b_by_id["bio2"].collected_at == 0
        # Cross-cutting: subjects must remain empty.
        assert a.subjects == []
        assert b.subjects == []


def test_empty_lists_emit_no_packets(tmp_path: Path) -> None:
    """Spec §5.4 step 5 says "zero or more"; empty subject/sample lists
    MUST NOT emit 0x19 or 0x1A packets on the wire.

    Uses a reference-only fixture to ensure the v0.11 prelude IS
    emitted (so we know we're seeing real prelude content), and
    verifies neither 0x19 nor 0x1A appears.
    """
    src = tmp_path / "src.tio"
    SpectralDataset.write_minimal(
        src,
        title="ref_only",
        isa_investigation_id="",
        runs={},
    )
    # Add a small reference to ensure v0.11 prelude triggers.
    from ttio.genomic.reference_import import ReferenceImport
    ref = ReferenceImport(
        uri="ref-only-fixture-v1",
        chromosomes=["chr1"],
        sequences=[b"ACGTACGT"],
    )
    with SpectralDataset.open(src, writable=True) as ds_w:
        ref.write_to_dataset(ds_w)

    out = io.BytesIO()
    with SpectralDataset.open(src) as ds:
        with TransportWriter(out) as w:
            w.write_dataset(ds)

    r = TransportReader(io.BytesIO(out.getvalue()))
    records = r.records_for_test()
    for rec in records:
        assert rec.header.packet_type != int(PacketType.SUBJECT_METADATA), (
            "ref-only fixture must not emit SUBJECT_METADATA"
        )
        assert rec.header.packet_type != int(PacketType.SAMPLE_METADATA), (
            "ref-only fixture must not emit SAMPLE_METADATA"
        )


def test_subjects_emit_before_samples_per_spec_5_4_3(tmp_path: Path) -> None:
    """Both populated: exactly one 0x19 followed by exactly one 0x1A
    (per spec §5.4 step 5, subjects-first ordering — so a downstream
    reader sees subjects ahead of any samples that soft-FK into them).
    """
    src = tmp_path / "both.tio"
    subjects = [
        Subject(external_id="S1", project="P", sex="F", birth_year=1980),
    ]
    samples = [
        Sample(
            sample_id="bio1",
            subject_external_id="S1",
            sample_kind="tissue",
            collected_at=1700000000,
        ),
    ]
    SpectralDataset.write_minimal(
        src,
        title="both",
        isa_investigation_id="",
        runs={},
        subjects=subjects,
        samples=samples,
    )

    out = io.BytesIO()
    with SpectralDataset.open(src) as ds:
        with TransportWriter(out) as w:
            w.write_dataset(ds)

    r = TransportReader(io.BytesIO(out.getvalue()))
    records = r.records_for_test()
    subj_idx = -1
    samp_idx = -1
    for i, rec in enumerate(records):
        t = rec.header.packet_type
        if t == int(PacketType.SUBJECT_METADATA):
            assert subj_idx == -1, "duplicate SUBJECT_METADATA"
            subj_idx = i
        elif t == int(PacketType.SAMPLE_METADATA):
            assert samp_idx == -1, "duplicate SAMPLE_METADATA"
            samp_idx = i
    assert subj_idx > 0, "expected exactly one SUBJECT_METADATA packet"
    assert samp_idx > 0, "expected exactly one SAMPLE_METADATA packet"
    assert subj_idx < samp_idx, (
        "per spec §5.4 step 5: SUBJECT_METADATA must precede SAMPLE_METADATA"
    )

File: objc/Tests/TestTransportSubjectsSamples.m
/*
 * TestTransportSubjectsSamples.m — Stage 6 (transport-spec v0.11,
 * Deferral 2 / Task 6.4 Commit 3).
 *
 * Exercises SUBJECT_METADATA (0x19) + SAMPLE_METADATA (0x1A) writer +
 * reader. Each packet carries a single length-prefixed Apache Arrow
 * IPC stream per transport-spec §4.22:
 *   uint32 arrow_ipc_length (LE)
 *   bytes  arrow_ipc[arrow_ipc_length]
 *
 * Round-trip path:
 *   writer -writeSubjectMetadata:/-writeSampleMetadata:
 *     -> Arrow IPC bytes
 *       -> reader 0x19/0x1A decoder
 *         -> /study/{subjects,samples}/<id>/ HDF5 per-row groups
 *           -> SpectralDataset.subjects / .samples accessors
 *
 * Cross-language parity:
 *   Java TransportSubjectsSamplesTest (commit dd211600)
 *   Python tests/test_transport_subjects_samples.py (commit 00c7e1b7)
 *
 * SPDX-License-Identifier: Apache-2.0
 */
#import <Foundation/Foundation.h>
#import "Testing.h"

#import "Transport/TTIOTransportPacket.h"
#import "Transport/TTIOTransportReader.h"
#import "Transport/TTIOTransportReader+Internal.h"
#import "Transport/TTIOTransportWriter.h"
#import "Dataset/TTIOSpectralDataset.h"
#import "Dataset/TTIOSubject.h"
#import "Dataset/TTIOSample.h"
#include <unistd.h>

// -------- helpers ----------------------------------------------------------

static NSString *makeTempPath_TSS(NSString *suffix)
{
    NSString *base = [NSString stringWithFormat:@"ttio_tr_ss_%d_%@.tio",
                       (int)getpid(), suffix];
    return [NSTemporaryDirectory()
        stringByAppendingPathComponent:base];
}

// ── 1. subjects-only round-trip via writeDataset: ────────────────────

static void testSubjectsOnlyRoundTrip(void)
{
    // Build a fixture .tio via SpectralDataset.writeMinimalToPath:
    // then layer subjects directly by reopening via the same path
    // the SpectralDataset Commit 1 tests use. Then drive a full
    // writeDataset: -> writeTtioToPath: round-trip and verify the
    // subjects accessor on the materialised file matches.
    TTIOSubject *s1 = [[TTIOSubject alloc] initWithExternalId:@"S1"
                                                         project:@"STUDY-A"
                                                             sex:@"F"
                                                       birthYear:1985
                                                      attributes:@{ @"site": @"NYC" }];
    TTIOSubject *s2 = [[TTIOSubject alloc] initWithExternalId:@"S2"
                                                         project:nil
                                                             sex:nil
                                                       birthYear:0
                                                      attributes:nil];

    // Write packets directly via the fine-grained writer API.
    NSMutableData *buf = [NSMutableData data];
    TTIOTransportWriter *w =
        [[TTIOTransportWriter alloc] initWithMutableData:buf];
    NSError *err = nil;
    PASS([w writeStreamHeaderWithFormatVersion:@"1.2"
                                          title:@"subjects-only"
                               isaInvestigation:@""
                                       features:@[TTIOTransportV011Feature]
                                      nDatasets:0
                                          error:&err],
         "6.4 subj-only: StreamHeader");
    PASS([w writeSubjectMetadata:(@[s1, s2]) error:&err],
         "6.4 subj-only: writeSubjectMetadata emitted");
    PASS([w writeEndOfStreamWithError:&err],
         "6.4 subj-only: EndOfStream");

    // Verify exactly one 0x19 and zero 0x1A.
    TTIOTransportReader *probe =
        [[TTIOTransportReader alloc] initWithData:buf];
    NSArray<TTIOTransportPacketRecord *> *records =
        [probe readAllPacketsWithError:&err];
    int subjCount = 0, sampCount = 0;
    for (TTIOTransportPacketRecord *rec in records) {
        if (rec.header.packetType
                == TTIOTransportPacketSubjectMetadata) subjCount++;
        if (rec.header.packetType
                == TTIOTransportPacketSampleMetadata)  sampCount++;
    }
    PASS(subjCount == 1, "6.4 subj-only: exactly one 0x19 on wire");
    PASS(sampCount == 0, "6.4 subj-only: zero 0x1A packets");

    NSString *outPath = makeTempPath_TSS(@"subj-rt");
    unlink([outPath fileSystemRepresentation]);
    TTIOTransportReader *r =
        [[TTIOTransportReader alloc] initWithData:buf];
    err = nil;
    PASS([r writeTtioToPath:outPath error:&err] && err == nil,
         "6.4 subj-only: writeTtioToPath materialised");

    TTIOSpectralDataset *back =
        [TTIOSpectralDataset readFromFilePath:outPath error:&err];
    PASS(back != nil, "6.4 subj-only: re-opened materialised .tio");
    PASS(back.subjects.count == 2, "6.4 subj-only: 2 subjects materialised");
    PASS(back.samples.count == 0,  "6.4 subj-only: no samples on disk");

    TTIOSubject *r1 = nil, *r2 = nil;
    for (TTIOSubject *s in back.subjects) {
        if ([s.externalId isEqualToString:@"S1"]) r1 = s;
        if ([s.externalId isEqualToString:@"S2"]) r2 = s;
    }
    PASS(r1 != nil && r2 != nil,
         "6.4 subj-only: both subjects present after wire round-trip");
    if (r1 != nil) {
        PASS([r1.project isEqualToString:@"STUDY-A"], "6.4 subj-only: r1.project");
        PASS([r1.sex isEqualToString:@"F"],            "6.4 subj-only: r1.sex");
        PASS(r1.birthYear == 1985,                      "6.4 subj-only: r1.birthYear");
        PASS([r1.attributes[@"site"] isEqualToString:@"NYC"],
             "6.4 subj-only: r1.attributes.site");
    }
    if (r2 != nil) {
        PASS([r2.project isEqualToString:@""],
             "6.4 subj-only: r2 empty project (null wire -> @\"\")");
        PASS(r2.birthYear == 0,
             "6.4 subj-only: r2 sentinel birthYear (null wire -> 0)");
    }

    [back closeFile];
    unlink([outPath fileSystemRepresentation]);
}

// ── 2. samples-only round-trip ───────────────────────────────────────

static void testSamplesOnlyRoundTrip(void)
{
    TTIOSample *m1 = [[TTIOSample alloc] initWithSampleId:@"M1"
                                         subjectExternalId:@"S1"
                                                sampleKind:@"plasma"
                                               collectedAt:1716671000
                                                attributes:@{ @"freezer": @"L-80" }];
    TTIOSample *m2 = [[TTIOSample alloc] initWithSampleId:@"M2"
                                         subjectExternalId:nil
                                                sampleKind:nil
                                               collectedAt:0
                                                attributes:nil];

    NSMutableData *buf = [NSMutableData data];
    TTIOTransportWriter *w =
        [[TTIOTransportWriter alloc] initWithMutableData:buf];
    NSError *err = nil;
    PASS([w writeStreamHeaderWithFormatVersion:@"1.2"
                                          title:@"samples-only"
                               isaInvestigation:@""
                                       features:@[TTIOTransportV011Feature]
                                      nDatasets:0
                                          error:&err],
         "6.4 samp-only: StreamHeader");
    PASS([w writeSampleMetadata:(@[m1, m2]) error:&err],
         "6.4 samp-only: writeSampleMetadata emitted");
    PASS([w writeEndOfStreamWithError:&err],
         "6.4 samp-only: EndOfStream");

    TTIOTransportReader *probe =
        [[TTIOTransportReader alloc] initWithData:buf];
    NSArray<TTIOTransportPacketRecord *> *records =
        [probe readAllPacketsWithError:&err];
    int subjCount = 0, sampCount = 0;
    for (TTIOTransportPacketRecord *rec in records) {
        if (rec.header.packetType
                == TTIOTransportPacketSubjectMetadata) subjCount++;
        if (rec.header.packetType
                == TTIOTransportPacketSampleMetadata)  sampCount++;
    }
    PASS(subjCount == 0, "6.4 samp-only: zero 0x19 packets");
    PASS(sampCount == 1, "6.4 samp-only: exactly one 0x1A on wire");

    NSString *outPath = makeTempPath_TSS(@"samp-rt");
    unlink([outPath fileSystemRepresentation]);
    TTIOTransportReader *r =
        [[TTIOTransportReader alloc] initWithData:buf];
    err = nil;
    PASS([r writeTtioToPath:outPath error:&err] && err == nil,
         "6.4 samp-only: writeTtioToPath materialised");

    TTIOSpectralDataset *back =
        [TTIOSpectralDataset readFromFilePath:outPath error:&err];
    PASS(back != nil, "6.4 samp-only: re-opened");
    PASS(back.subjects.count == 0, "6.4 samp-only: no subjects on disk");
    PASS(back.samples.count == 2,  "6.4 samp-only: 2 samples materialised");

    TTIOSample *r1 = nil, *r2 = nil;
    for (TTIOSample *s in back.samples) {
        if ([s.sampleId isEqualToString:@"M1"]) r1 = s;
        if ([s.sampleId isEqualToString:@"M2"]) r2 = s;
    }
    PASS(r1 != nil && r2 != nil,
         "6.4 samp-only: both samples present");
    if (r1 != nil) {
        PASS([r1.subjectExternalId isEqualToString:@"S1"],
             "6.4 samp-only: r1.subjectExternalId");
        PASS([r1.sampleKind isEqualToString:@"plasma"],
             "6.4 samp-only: r1.sampleKind");
        PASS(r1.collectedAt == 1716671000,
             "6.4 samp-only: r1.collectedAt");
        PASS([r1.attributes[@"freezer"] isEqualToString:@"L-80"],
             "6.4 samp-only: r1.attributes.freezer");
    }
    if (r2 != nil) {
        PASS([r2.subjectExternalId isEqualToString:@""],
             "6.4 samp-only: r2 empty FK (null wire -> @\"\")");
        PASS([r2.sampleKind isEqualToString:@""],
             "6.4 samp-only: r2 empty kind");
        PASS(r2.collectedAt == 0,
             "6.4 samp-only: r2 sentinel collectedAt");
    }

    [back closeFile];
    unlink([outPath fileSystemRepresentation]);
}

// ── 3. empty subjects/samples -> no packets ──────────────────────────

static void testEmptyEmitsNoPackets(void)
{
    NSMutableData *buf = [NSMutableData data];
    TTIOTransportWriter *w =
        [[TTIOTransportWriter alloc] initWithMutableData:buf];
    NSError *err = nil;
    PASS([w writeStreamHeaderWithFormatVersion:@"1.2"
                                          title:@"empty"
                               isaInvestigation:@""
                                       features:@[]
                                      nDatasets:0
                                          error:&err],
         "6.4 empty: StreamHeader");
    // Empty lists must produce zero packets per spec §5.4 step 5.
    PASS([w writeSubjectMetadata:@[] error:&err],
         "6.4 empty: writeSubjectMetadata @[] no-op");
    PASS([w writeSampleMetadata:@[] error:&err],
         "6.4 empty: writeSampleMetadata @[] no-op");
    PASS([w writeEndOfStreamWithError:&err],
         "6.4 empty: EndOfStream");

    TTIOTransportReader *probe =
        [[TTIOTransportReader alloc] initWithData:buf];
    NSArray<TTIOTransportPacketRecord *> *records =
        [probe readAllPacketsWithError:&err];
    int subjCount = 0, sampCount = 0;
    for (TTIOTransportPacketRecord *rec in records) {
        if (rec.header.packetType
                == TTIOTransportPacketSubjectMetadata) subjCount++;
        if (rec.header.packetType
                == TTIOTransportPacketSampleMetadata)  sampCount++;
    }
    PASS(subjCount == 0,
         "6.4 empty: writeSubjectMetadata @[] emits zero 0x19 packets");
    PASS(sampCount == 0,
         "6.4 empty: writeSampleMetadata @[] emits zero 0x1A packets");
}

// ── 4. both populated -> 0x19 emitted before 0x1A in §5.4 order ──────
//
// Build a source .tio carrying both subjects + samples (via the same
// build-fixture pattern Commit 1 uses), drive it through
// -writeDataset:, then assert the on-wire ordering: 0x19 must precede
// 0x1A so forward references resolve during streaming.

#import "HDF5/TTIOHDF5File.h"
#import "HDF5/TTIOHDF5Group.h"

static NSString *buildSourceTioWithBoth(TTIOSubject *s,
                                         TTIOSample *m,
                                         NSError **error)
{
    NSString *path = makeTempPath_TSS(@"both-src");
    unlink([path fileSystemRepresentation]);
    BOOL ok = [TTIOSpectralDataset writeMinimalToPath:path
                                                  title:@"both-src"
                                     isaInvestigationId:@""
                                                 msRuns:@{}
                                        identifications:nil
                                        quantifications:nil
                                      provenanceRecords:nil
                                                  error:error];
    if (!ok) return nil;
    TTIOHDF5File *f = [TTIOHDF5File openAtPath:path error:error];
    if (f == nil) return nil;
    TTIOHDF5Group *root = [f rootGroup];
    TTIOHDF5Group *study = [root openGroupNamed:@"study" error:error];
    if (study == nil) { [f close]; return nil; }
    TTIOHDF5Group *sg = [study createGroupNamed:@"subjects" error:error];
    TTIOHDF5Group *srow = [sg createGroupNamed:s.externalId error:error];
    [srow setStringAttribute:@"external_id" value:s.externalId error:error];
    if (s.project.length > 0)
        [srow setStringAttribute:@"project" value:s.project error:error];
    if (s.sex.length > 0)
        [srow setStringAttribute:@"sex" value:s.sex error:error];
    [srow setIntegerAttribute:@"birth_year" value:s.birthYear error:error];
    [srow setStringAttribute:@"attributes_json" value:[s attributesJson] error:error];

    TTIOHDF5Group *smg = [study createGroupNamed:@"samples" error:error];
    TTIOHDF5Group *mrow = [smg createGroupNamed:m.sampleId error:error];
    [mrow setStringAttribute:@"sample_id" value:m.sampleId error:error];
    if (m.subjectExternalId.length > 0)
        [mrow setStringAttribute:@"subject_external_id" value:m.subjectExternalId error:error];
    if (m.sampleKind.length > 0)
        [mrow setStringAttribute:@"sample_kind" value:m.sampleKind error:error];
    [mrow setIntegerAttribute:@"collected_at" value:m.collectedAt error:error];
    [mrow setStringAttribute:@"attributes_json" value:[m attributesJson] error:error];
    [f close];
    return path;
}

static void testBothOrdering(void)
{
    TTIOSubject *s = [[TTIOSubject alloc] initWithExternalId:@"S1"
                                                        project:nil
                                                            sex:nil
                                                      birthYear:0
                                                     attributes:nil];
    TTIOSample *m = [[TTIOSample alloc] initWithSampleId:@"M1"
                                        subjectExternalId:@"S1"
                                               sampleKind:nil
                                              collectedAt:0
                                               attributes:nil];
    NSError *err = nil;
    NSString *srcPath = buildSourceTioWithBoth(s, m, &err);
    PASS(srcPath != nil, "6.4 both: source fixture built");

    TTIOSpectralDataset *src =
        [TTIOSpectralDataset readFromFilePath:srcPath error:&err];
    PASS(src != nil, "6.4 both: source dataset opens");
    PASS(src.subjects.count == 1, "6.4 both: src subjects count");
    PASS(src.samples.count  == 1, "6.4 both: src samples count");

    NSMutableData *buf = [NSMutableData data];
    TTIOTransportWriter *w =
        [[TTIOTransportWriter alloc] initWithMutableData:buf];
    PASS([w writeDataset:src error:&err],
         "6.4 both: writeDataset emitted full prelude");

    // Walk packet sequence; assert 0x19 appears before 0x1A.
    TTIOTransportReader *probe =
        [[TTIOTransportReader alloc] initWithData:buf];
    NSArray<TTIOTransportPacketRecord *> *records =
        [probe readAllPacketsWithError:&err];
    NSInteger firstSubj = -1, firstSamp = -1;
    NSInteger idx = 0;
    for (TTIOTransportPacketRecord *rec in records) {
        if (rec.header.packetType == TTIOTransportPacketSubjectMetadata
            && firstSubj < 0) firstSubj = idx;
        if (rec.header.packetType == TTIOTransportPacketSampleMetadata
            && firstSamp < 0) firstSamp = idx;
        idx++;
    }
    PASS(firstSubj >= 0, "6.4 both: 0x19 present in stream");
    PASS(firstSamp >= 0, "6.4 both: 0x1A present in stream");
    PASS(firstSubj < firstSamp,
         "6.4 both: 0x19 emitted BEFORE 0x1A (§5.4 ordering)");

    // Round-trip to a fresh .tio and verify both surfaced.
    NSString *outPath = makeTempPath_TSS(@"both-rt");
    unlink([outPath fileSystemRepresentation]);
    TTIOTransportReader *r =
        [[TTIOTransportReader alloc] initWithData:buf];
    PASS([r writeTtioToPath:outPath error:&err],
         "6.4 both: writeTtioToPath materialised");
    TTIOSpectralDataset *back =
        [TTIOSpectralDataset readFromFilePath:outPath error:&err];
    PASS(back != nil && back.subjects.count == 1 && back.samples.count == 1,
         "6.4 both: both accessors round-tripped");

    [back closeFile];
    [src closeFile];
    unlink([outPath fileSystemRepresentation]);
    unlink([srcPath fileSystemRepresentation]);
}

void testTransportSubjectsSamples(void)
{
    testSubjectsOnlyRoundTrip();
    testSamplesOnlyRoundTrip();
    testEmptyEmitsNoPackets();
    testBothOrdering();
}
